File: crates/tui/extension-host/src/protocol.ts
/**
 * Codewhale extension-host protocol, version 1 (phase 1 subset).
 *
 * The Rust serde types in `crates/tui/src/extension_host/protocol.rs` are the
 * source of truth. Phase 1 keeps this file hand-written and pins both sides
 * to the shared corpus in `crates/tui/tests/fixtures/extension_host/protocol`,
 * which both parsers must accept (or reject) identically and round-trip.
 *
 * Frame: 4-byte magic `CWX1`, u32 little-endian payload length, UTF-8 JSON.
 * Envelope: JSON-RPC 2.0.
 */

export const PROTOCOL_VERSION = 1
export const MAGIC = Buffer.from('CWX1', 'ascii')
export const MAX_FRAME = 32 * 1024 * 1024
export const HEADER_LEN = 8

/** JSON-RPC error codes used on this channel. */
export const ErrorCode = {
  ParseError: -32700,
  InvalidRequest: -32600,
  MethodNotFound: -32601,
  InvalidParams: -32602,
  Internal: -32603,
  /** The tool body threw. */
  ExecutionFailed: -32000,
  /** The handle is unknown, revoked, or its owner is not active. */
  NotAvailable: -32001,
  /** Cancelled by `$/cancel` (LSP's RequestCancelled). */
  Cancelled: -32800,
} as const

/** Methods the core sends to the host. */
export const CORE_TO_HOST = {
  requests: ['host/initialize', 'host/shutdown', 'ext/activate', 'ext/deactivate', 'tool/call'],
  notifications: ['$/cancel'],
} as const

/** Methods the host sends to the core. */
export const HOST_TO_CORE = {
  requests: ['registry/register', 'registry/unregister'],
  notifications: ['host/hello', 'host/ready', 'ext/faulted', 'log', '$/cancel'],
} as const

export type Json = null | boolean | number | string | Json[] | { [key: string]: Json }

export interface OwnerRef {
  plugin_id: string
  generation: number
  owner_token: string
}

export interface ToolSpecWire {
  name: string
  description: string
  input_schema: { [key: string]: Json }
}

export interface ContentBlockWire {
  type: 'text'
  text: string
}

export interface ToolResultWire {
  content: ContentBlockWire[]
  is_error: boolean
  structured?: Json
}

export interface RpcErrorWire {
  code: number
  message: string
  data?: Json
}

export type Message =
  | { jsonrpc: '2.0'; id: number; method: string; params?: any }
  | { jsonrpc: '2.0'; method: string; params?: any }
  | { jsonrpc: '2.0'; id: number; result: any }
  | { jsonrpc: '2.0'; id: number; error: RpcErrorWire }

export class FrameError extends Error {
  constructor(message: string) {
    super(message)
    this.name = 'FrameError'
  }
}

export class ProtocolError extends Error {
  constructor(message: string) {
    super(message)
    this.name = 'ProtocolError'
  }
}

/** Encode one message into a `CWX1` frame. Oversized payloads are refused, never truncated. */
export function encodeFrame(message: unknown): Buffer {
  const payload = Buffer.from(JSON.stringify(message), 'utf8')
  if (payload.length > MAX_FRAME) {
    throw new FrameError(`frame of ${payload.length} bytes exceeds MAX_FRAME ${MAX_FRAME}`)
  }
  const header = Buffer.alloc(HEADER_LEN)
  MAGIC.copy(header, 0)
  header.writeUInt32LE(payload.length, 4)
  return Buffer.concat([header, payload])
}

/** Incremental `CWX1` decoder. Throws `FrameError` on bad magic, bad length, or bad JSON. */
export class FrameDecoder {
  private buffer: Buffer = Buffer.alloc(0)

  push(chunk: Buffer): unknown[] {
    this.buffer = this.buffer.length === 0 ? chunk : Buffer.concat([this.buffer, chunk])
    const out: unknown[] = []
    while (this.buffer.length >= HEADER_LEN) {
      if (!this.buffer.subarray(0, 4).equals(MAGIC)) {
        throw new FrameError('bad frame magic')
      }
      const length = this.buffer.readUInt32LE(4)
      if (length > MAX_FRAME) throw new FrameError(`frame length ${length} exceeds MAX_FRAME`)
      if (this.buffer.length < HEADER_LEN + length) break
      const payload = this.buffer.subarray(HEADER_LEN, HEADER_LEN + length)
      this.buffer = this.buffer.subarray(HEADER_LEN + length)
      let value: unknown
      try {
        value = JSON.parse(payload.toString('utf8'))
      } catch (error) {
        throw new FrameError(`frame payload is not JSON: ${(error as Error).message}`)
      }
      out.push(value)
    }
    return out
  }
}

// ---------------------------------------------------------------------------
// Validation. Host→core output is validated strictly before it is sent (the
// Rust side uses deny_unknown_fields for it); core→host input is validated
// tolerantly (unknown fields ignored), mirroring the Rust side.
// ---------------------------------------------------------------------------

type Shape = Record<string, 'string' | 'number' | 'boolean' | 'object' | 'array' | 'json' | 'owner' | 'u64'>

function isObject(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null && !Array.isArray(value)
}

function checkShape(
  where: string,
  value: unknown,
  required: Shape,
  optional: Shape = {},
  strict = false,
): asserts value is Record<string, any> {
  if (!isObject(value)) throw new ProtocolError(`${where}: expected an object`)
  for (const [key, kind] of Object.entries(required)) {
    if (!(key in value)) throw new ProtocolError(`${where}: missing field \`${key}\``)
    checkKind(`${where}.${key}`, value[key], kind)
  }
  for (const [key, kind] of Object.entries(optional)) {
    if (key in value && value[key] !== undefined) checkKind(`${where}.${key}`, value[key], kind)
  }
  if (strict) {
    for (const key of Object.keys(value)) {
      if (!(key in required) && !(key in optional)) {
        throw new ProtocolError(`${where}: unknown field \`${key}\``)
      }
    }
  }
}

function checkKind(where: string, value: unknown, kind: Shape[string]) {
  switch (kind) {
    case 'string':
      if (typeof value !== 'string') throw new ProtocolError(`${where}: expected a string`)
      return
    case 'number':
      if (typeof value !== 'number' || !Number.isFinite(value)) throw new ProtocolError(`${where}: expected a number`)
      return
    case 'u64':
      if (typeof value !== 'number' || !Number.isSafeInteger(value) || value < 0) {
        throw new ProtocolError(`${where}: expected an unsigned integer`)
      }
      return
    case 'boolean':
      if (typeof value !== 'boolean') throw new ProtocolError(`${where}: expected a boolean`)
      return
    case 'object':
      if (!isObject(value)) throw new ProtocolError(`${where}: expected an object`)
      return
    case 'array':
      if (!Array.isArray(value)) throw new ProtocolError(`${where}: expected an array`)
      return
    case 'owner':
      checkShape(where, value, { plugin_id: 'string', generation: 'u64', owner_token: 'string' }, {}, true)
      return
    case 'json':
      return
  }
}

/** Params validation per method and direction. `strict` = deny unknown fields. */
const PARAMS: Record<string, { dir: 'core' | 'host'; required: Shape; optional?: Shape; check?: (p: any, strict: boolean) => void }> = {
  // host → core
  'host/hello': {
    dir: 'host',
    required: { protocol: 'object', host_version: 'string', bundle_sha256: 'string', node_version: 'string' },
    check: (p, strict) => checkShape('host/hello.protocol', p.protocol, { min: 'u64', max: 'u64' }, {}, strict),
  },
  'host/ready': { dir: 'host', required: {} },
  'registry/register': {
    dir: 'host',
    required: { owner: 'owner', kind: 'string', spec: 'object' },
    check: (p, strict) => {
      if (p.kind !== 'tool') throw new ProtocolError(`registry/register: unsupported kind \`${p.kind}\``)
      checkShape('registry/register.spec', p.spec, { name: 'string', description: 'string', input_schema: 'object' }, {}, strict)
    },
  },
  'registry/unregister': { dir: 'host', required: { owner: 'owner', handle: 'u64' } },
  'ext/faulted': { dir: 'host', required: { owner: 'owner', error: 'string' } },
  log: { dir: 'host', required: { level: 'string', msg: 'string' }, optional: { plugin_id: 'string' } },
  // core → host
  'host/initialize': {
    dir: 'core',
    required: { protocol: 'u64', limits: 'object' },
    check: (p, strict) =>
      checkShape(
        'host/initialize.limits',
        p.limits,
        { max_frame: 'u64', max_inflight: 'u64', dispose_deadline_ms: 'u64', activate_deadline_ms: 'u64' },
        {},
        strict,
      ),
  },
  'host/shutdown': { dir: 'core', required: {} },
  'ext/activate': {
    dir: 'core',
    required: { owner: 'owner', plugin_name: 'string', entry: 'object' },
    optional: { config: 'json' },
    check: (p, strict) => checkShape('ext/activate.entry', p.entry, { path: 'string', sha256: 'string' }, {}, strict),
  },
  'ext/deactivate': { dir: 'core', required: { owner: 'owner' } },
  'tool/call': {
    dir: 'core',
    required: { handle: 'u64', call_id: 'string', input: 'json', deadline_ms: 'u64' },
  },
}

/** `$/cancel` flows both ways. */
const CANCEL_SHAPE: Shape = { id: 'u64' }

export type Direction = 'core_to_host' | 'host_to_core'

/**
 * Validate one decoded message travelling in `direction`. Host→core messages
 * are validated strictly (unknown fields rejected); core→host tolerantly.
 * Responses are validated as envelopes only: their result shape depends on the
 * request, which the RPC layer checks.
 */
export function validateMessage(value: unknown, direction: Direction): Message {
  const strict = direction === 'host_to_core'
  if (!isObject(value)) throw new ProtocolError('message: expected an object')
  if (value.jsonrpc !== '2.0') throw new ProtocolError('message: jsonrpc must be "2.0"')
  const hasId = 'id' in value
  if (hasId) checkKind('message.id', value.id, 'u64')
  if ('method' in value) {
    checkShape('message', value, { jsonrpc: 'string', method: 'string' }, { id: 'u64', params: 'json' }, strict)
    const method = value.method as string
    const params = value.params ?? {}
    if (method === '$/cancel') {
      if (hasId) throw new ProtocolError('$/cancel must be a notification')
      checkShape('$/cancel', params, CANCEL_SHAPE, {}, strict)
      return value as Message
    }
    const spec = PARAMS[method]
    const expectedDir = direction === 'host_to_core' ? 'host' : 'core'
    if (!spec || spec.dir !== expectedDir) {
      throw new ProtocolError(`unknown ${direction} method \`${method}\``)
    }
    const isRequest = (direction === 'host_to_core' ? HOST_TO_CORE.requests : CORE_TO_HOST.requests).includes(
      method as never,
    )
    if (isRequest !== hasId) {
      throw new ProtocolError(`\`${method}\` must be ${isRequest ? 'a request (with id)' : 'a notification (no id)'}`)
    }
    checkShape(method, params, spec.required, spec.optional ?? {}, strict)
    spec.check?.(params, strict)
    return value as Message
  }
  if (!hasId) throw new ProtocolError('response: missing id')
  if ('error' in value) {
    checkShape('message', value, { jsonrpc: 'string', id: 'u64', error: 'object' }, {}, strict)
    checkShape('error', value.error, { code: 'number', message: 'string' }, { data: 'json' }, strict)
    return value as Message
  }
  if (!('result' in value)) throw new ProtocolError('response: needs result or error')
  checkShape('message', value, { jsonrpc: 'string', id: 'u64', result: 'json' }, {}, strict)
  return value as Message
}

File: crates/tui/src/extension_host/protocol.rs
//! Extension-host protocol v1 — the phase-1 subset, and its source of truth.
//!
//! Frame: 4-byte magic `CWX1`, u32 little-endian payload length, UTF-8 JSON.
//! Envelope: JSON-RPC 2.0. A length prefix (not NDJSON) makes a plugin that
//! writes raw bytes to the channel detectable: bad magic or length ends the
//! host instead of desynchronising it.
//!
//! Host→core types use `deny_unknown_fields` — host output is untrusted
//! input. Core→host types are what this side writes. The TypeScript mirror is
//! hand-written in phase 1 (`crates/tui/extension-host/src/protocol.ts`); both
//! parse the shared corpus in `tests/fixtures/extension_host/protocol`.
//!
//! There is deliberately no method that expresses approval, and nothing a
//! host can send makes the core *do* anything in phase 1: registrations are
//! admitted or refused, and tool calls only flow core→host after the gate.

use serde::de::DeserializeOwned;
use serde::{Deserialize, Deserializer, Serialize};
use serde_json::{Map, Value, json};
use tokio::io::{AsyncRead, AsyncReadExt};

pub const PROTOCOL_VERSION: u32 = 1;
pub const MAGIC: [u8; 4] = *b"CWX1";
pub const HEADER_LEN: usize = 8;
/// Enough for base64 screenshots later; anything larger is refused, never truncated.
pub const MAX_FRAME: usize = 32 * 1024 * 1024;
/// Requests in flight per direction.
pub const MAX_INFLIGHT: usize = 256;

/// JSON-RPC error codes used on this channel.
pub mod error_code {
    pub const INVALID_PARAMS: i64 = -32602;
    /// Unknown, revoked, or not-yet-active handle.
    pub const NOT_AVAILABLE: i64 = -32001;
    /// Cancelled by `$/cancel`.
    pub const CANCELLED: i64 = -32800;
}

#[derive(Debug, thiserror::Error)]
pub enum FrameError {
    #[error("bad frame magic")]
    BadMagic,
    #[error("frame of {0} bytes exceeds MAX_FRAME")]
    TooLarge(usize),
    #[error("frame payload is not JSON: {0}")]
    Json(#[from] serde_json::Error),
    #[error("channel read failed: {0}")]
    Io(#[from] std::io::Error),
}

#[derive(Debug, Clone, PartialEq, Eq, thiserror::Error)]
#[error("{0}")]
pub struct ProtocolError(pub String);

fn perr(message: impl Into<String>) -> ProtocolError {
    ProtocolError(message.into())
}

/// Encode one message into a `CWX1` frame.
pub fn encode_frame(message: &Value) -> Result<Vec<u8>, FrameError> {
    let payload = serde_json::to_vec(message)?;
    if payload.len() > MAX_FRAME {
        return Err(FrameError::TooLarge(payload.len()));
    }
    let mut frame = Vec::with_capacity(HEADER_LEN + payload.len());
    frame.extend_from_slice(&MAGIC);
    frame.extend_from_slice(&(payload.len() as u32).to_le_bytes());
    frame.extend_from_slice(&payload);
    Ok(frame)
}

/// Read one frame. `Ok(None)` is a clean EOF at a frame boundary. The payload
/// buffer is allocated only after the length is checked against `MAX_FRAME`.
pub async fn read_frame<R: AsyncRead + Unpin>(reader: &mut R) -> Result<Option<Value>, FrameError> {
    let mut header = [0u8; HEADER_LEN];
    let mut filled = 0;
    while filled < HEADER_LEN {
        let read = reader.read(&mut header[filled..]).await?;
        if read == 0 {
            if filled == 0 {
                return Ok(None);
            }
            return Err(FrameError::Io(std::io::ErrorKind::UnexpectedEof.into()));
        }
        filled += read;
    }
    if header[..4] != MAGIC {
        return Err(FrameError::BadMagic);
    }
    let length = u32::from_le_bytes([header[4], header[5], header[6], header[7]]) as usize;
    if length > MAX_FRAME {
        return Err(FrameError::TooLarge(length));
    }
    let mut payload = vec![0u8; length];
    reader.read_exact(&mut payload).await?;
    Ok(Some(serde_json::from_slice(&payload)?))
}

/// `Some(value)` even for an explicit JSON `null`, so `"result": null` is a
/// present result and round-trips.
fn present<'de, D: Deserializer<'de>>(deserializer: D) -> Result<Option<Value>, D::Error> {
    Value::deserialize(deserializer).map(Some)
}

// ---------------------------------------------------------------------------
// Shared types
// ---------------------------------------------------------------------------

/// Identity of one activation. `generation` bumps on every (re)activation;
/// `owner_token` is 128+ random bits minted by the core and handed over only
/// in `ext/activate`. It catches bugs and stale fibers; it is not a boundary
/// against a malicious plugin in the same process (see the design, §4.4).
#[derive(Debug, Clone, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct OwnerRef {
    pub plugin_id: String,
    pub generation: u64,
    pub owner_token: String,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct RpcErrorWire {
    pub code: i64,
    pub message: String,
    #[serde(
        default,
        skip_serializing_if = "Option::is_none",
        deserialize_with = "present"
    )]
    pub data: Option<Value>,
}

// ---------------------------------------------------------------------------
// Host → core (strict)
// ---------------------------------------------------------------------------

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct ProtocolRange {
    pub min: u32,
    pub max: u32,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct HelloParams {
    pub protocol: ProtocolRange,
    pub host_version: String,
    pub bundle_sha256: String,
    pub node_version: String,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct EmptyParams {}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum RegisterKind {
    /// The only kind in phase 1.
    Tool,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct ToolSpecWire {
    pub name: String,
    pub description: String,
    pub input_schema: Map<String, Value>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct RegisterParams {
    pub owner: OwnerRef,
    pub kind: RegisterKind,
    pub spec: ToolSpecWire,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct UnregisterParams {
    pub owner: OwnerRef,
    pub handle: u64,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct FaultedParams {
    pub owner: OwnerRef,
    pub error: String,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct LogParams {
    pub level: String,
    pub msg: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub plugin_id: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct CancelParams {
    pub id: u64,
}

#[derive(Debug, Clone, PartialEq)]
pub enum HostRequest {
    Register(RegisterParams),
    Unregister(UnregisterParams),
}

#[derive(Debug, Clone, PartialEq)]
pub enum HostNotification {
    Hello(HelloParams),
    Ready,
    Faulted(FaultedParams),
    Log(LogParams),
    Cancel(CancelParams),
}

/// One decoded, validated host→core message.
#[derive(Debug, Clone, PartialEq)]
pub enum HostMessage {
    Request {
        id: u64,
        request: HostRequest,
    },
    Notification(HostNotification),
    Response {
        id: u64,
        outcome: Result<Value, RpcErrorWire>,
    },
}

#[derive(Deserialize)]
#[serde(deny_unknown_fields)]
struct Envelope {
    jsonrpc: String,
    #[serde(default)]
    id: Option<u64>,
    #[serde(default)]
    method: Option<String>,
    #[serde(default, deserialize_with = "present")]
    params: Option<Value>,
    #[serde(default, deserialize_with = "present")]
    result: Option<Value>,
    #[serde(default)]
    error: Option<RpcErrorWire>,
}

fn decode_envelope(value: Value) -> Result<Envelope, ProtocolError> {
    let envelope: Envelope =
        serde_json::from_value(value).map_err(|e| perr(format!("envelope: {e}")))?;
    if envelope.jsonrpc != "2.0" {
        return Err(perr("envelope: jsonrpc must be \"2.0\""));
    }
    Ok(envelope)
}

fn params<T: DeserializeOwned>(method: &str, params: Option<Value>) -> Result<T, ProtocolError> {
    serde_json::from_value(params.unwrap_or_else(|| json!({})))
        .map_err(|e| perr(format!("{method}: {e}")))
}

fn expect_request(method: &str, id: Option<u64>) -> Result<u64, ProtocolError> {
    id.ok_or_else(|| perr(format!("`{method}` must be a request (with id)")))
}

fn expect_notification(method: &str, id: Option<u64>) -> Result<(), ProtocolError> {
    match id {
        None => Ok(()),
        Some(_) => Err(perr(format!("`{method}` must be a notification (no id)"))),
    }
}

fn decode_response(
    envelope: Envelope,
) -> Result<(u64, Result<Value, RpcErrorWire>), ProtocolError> {
    let id = envelope.id.ok_or_else(|| perr("response: missing id"))?;
    if envelope.params.is_some() {
        return Err(perr("response: unexpected params"));
    }
    match (envelope.result, envelope.error) {
        (Some(result), None) => Ok((id, Ok(result))),
        (None, Some(error)) => Ok((id, Err(error))),
        _ => Err(perr("response: needs exactly one of result or error")),
    }
}

/// Parse and strictly validate one host→core message.
pub fn parse_host_message(value: Value) -> Result<HostMessage, ProtocolError> {
    let envelope = decode_envelope(value)?;
    let Some(method) = envelope.method.clone() else {
        let (id, outcome) = decode_response(envelope)?;
        return Ok(HostMessage::Response { id, outcome });
    };
    if envelope.result.is_some() || envelope.error.is_some() {
        return Err(perr(format!(
            "`{method}`: a request carries no result or error"
        )));
    }
    let id = envelope.id;
    let p = envelope.params;
    let message = match method.as_str() {
        "registry/register" => HostMessage::Request {
            id: expect_request(&method, id)?,
            request: HostRequest::Register(params(&method, p)?),
        },
        "registry/unregister" => HostMessage::Request {
            id: expect_request(&method, id)?,
            request: HostRequest::Unregister(params(&method, p)?),
        },
        "host/hello" => {
            expect_notification(&method, id)?;
            HostMessage::Notification(HostNotification::Hello(params(&method, p)?))
        }
        "host/ready" => {
            expect_notification(&method, id)?;
            let _: EmptyParams = params(&method, p)?;
            HostMessage::Notification(HostNotification::Ready)
        }
        "ext/faulted" => {
            expect_notification(&method, id)?;
            HostMessage::Notification(HostNotification::Faulted(params(&method, p)?))
        }
        "log" => {
            expect_notification(&method, id)?;
            HostMessage::Notification(HostNotification::Log(params(&method, p)?))
        }
        "$/cancel" => {
            expect_notification(&method, id)?;
            HostMessage::Notification(HostNotification::Cancel(params(&method, p)?))
        }
        other => return Err(perr(format!("unknown host_to_core method `{other}`"))),
    };
    Ok(message)
}

fn request_value(id: u64, method: &str, params: Value) -> Value {
    json!({"jsonrpc": "2.0", "id": id, "method": method, "params": params})
}

fn notification_value(method: &str, params: Value) -> Value {
    json!({"jsonrpc": "2.0", "method": method, "params": params})
}

fn response_value(id: u64, outcome: &Result<Value, RpcErrorWire>) -> Value {
    match outcome {
        Ok(result) => json!({"jsonrpc": "2.0", "id": id, "result": result}),
        Err(error) => json!({"jsonrpc": "2.0", "id": id, "error": error}),
    }
}

fn to_value<T: Serialize>(value: &T) -> Value {
    serde_json::to_value(value).expect("protocol types serialize")
}

#[cfg(test)]
impl HostMessage {
    /// Re-encode (used by the conformance corpus round-trip).
    #[must_use]
    pub fn to_value(&self) -> Value {
        match self {
            Self::Request { id, request } => match request {
                HostRequest::Register(p) => request_value(*id, "registry/register", to_value(p)),
                HostRequest::Unregister(p) => {
                    request_value(*id, "registry/unregister", to_value(p))
                }
            },
            Self::Notification(notification) => match notification {
                HostNotification::Hello(p) => notification_value("host/hello", to_value(p)),
                HostNotification::Ready => notification_value("host/ready", json!({})),
                HostNotification::Faulted(p) => notification_value("ext/faulted", to_value(p)),
                HostNotification::Log(p) => notification_value("log", to_value(p)),
                HostNotification::Cancel(p) => notification_value("$/cancel", to_value(p)),
            },
            Self::Response { id, outcome } => response_value(*id, outcome),
        }
    }
}

// ---------------------------------------------------------------------------
// Core → host (written by this side)
// ---------------------------------------------------------------------------

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct HostLimits {
    pub max_frame: u64,
    pub max_inflight: u64,
    pub dispose_deadline_ms: u64,
    pub activate_deadline_ms: u64,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct InitializeParams {
    pub protocol: u32,
    pub limits: HostLimits,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct EntryRef {
    pub path: String,
    pub sha256: String,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct ActivateParams {
    pub owner: OwnerRef,
    pub plugin_name: String,
    pub entry: EntryRef,
    #[serde(default = "empty_object")]
    pub config: Value,
}

fn empty_object() -> Value {
    json!({})
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct DeactivateParams {
    pub owner: OwnerRef,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct ToolCallParams {
    pub handle: u64,
    pub call_id: String,
    pub input: Value,
    pub deadline_ms: u64,
}

#[derive(Debug, Clone, PartialEq)]
pub enum CoreRequest {
    Initialize(InitializeParams),
    /// Production relies on stdin EOF at process exit (the host is shared by
    /// every engine in the process); the bounded shutdown is test-driven.
    #[cfg(test)]
    Shutdown,
    Activate(ActivateParams),
    Deactivate(DeactivateParams),
    ToolCall(ToolCallParams),
}

impl CoreRequest {
    #[must_use]
    pub fn method(&self) -> &'static str {
        match self {
            Self::Initialize(_) => "host/initialize",
            #[cfg(test)]
            Self::Shutdown => "host/shutdown",
            Self::Activate(_) => "ext/activate",
            Self::Deactivate(_) => "ext/deactivate",
            Self::ToolCall(_) => "tool/call",
        }
    }

    #[must_use]
    pub fn params(&self) -> Value {
        match self {
            Self::Initialize(p) => to_value(p),
            #[cfg(test)]
            Self::Shutdown => json!({}),
            Self::Activate(p) => to_value(p),
            Self::Deactivate(p) => to_value(p),
            Self::ToolCall(p) => to_value(p),
        }
    }

    #[must_use]
    pub fn to_value(&self, id: u64) -> Value {
        request_value(id, self.method(), self.params())
    }
}

/// `registry/register` answer: a handle, or a refusal the host reports as a
/// failed activation.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(untagged)]
pub enum RegisterResult {
    Admitted { handle: u64 },
    Refused { refused: String },
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(tag = "status", rename_all = "snake_case")]
pub enum ActivateResult {
    Ok { tools: Vec<String> },
    Failed { diagnostic: String },
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct DeactivateResult {
    pub disposed: bool,
    pub leaked: Vec<String>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum ContentBlockWire {
    Text { text: String },
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct ToolResultWire {
    pub content: Vec<ContentBlockWire>,
    pub is_error: bool,
    #[serde(
        default,
        skip_serializing_if = "Option::is_none",
        deserialize_with = "present"
    )]
    pub structured: Option<Value>,
}

/// One core→host message, parsed back (tests and the corpus only).
#[cfg(test)]
#[derive(Debug, Clone, PartialEq)]
pub enum CoreMessage {
    Request {
        id: u64,
        request: CoreRequest,
    },
    Cancel(CancelParams),
    Response {
        id: u64,
        outcome: Result<Value, RpcErrorWire>,
    },
}

/// Parse a core→host message (the shape this side writes).
#[cfg(test)]
pub fn parse_core_message(value: Value) -> Result<CoreMessage, ProtocolError> {
    let envelope = decode_envelope(value)?;
    let Some(method) = envelope.method.clone() else {
        let (id, outcome) = decode_response(envelope)?;
        return Ok(CoreMessage::Response { id, outcome });
    };
    let id = envelope.id;
    let p = envelope.params;
    if method == "$/cancel" {
        expect_notification(&method, id)?;
        return Ok(CoreMessage::Cancel(params(&method, p)?));
    }
    let id = expect_request(&method, id)?;
    let request = match method.as_str() {
        "host/initialize" => CoreRequest::Initialize(params(&method, p)?),
        "host/shutdown" => CoreRequest::Shutdown,
        "ext/activate" => CoreRequest::Activate(params(&method, p)?),
        "ext/deactivate" => CoreRequest::Deactivate(params(&method, p)?),
        "tool/call" => CoreRequest::ToolCall(params(&method, p)?),
        other => return Err(perr(format!("unknown core_to_host method `{other}`"))),
    };
    Ok(CoreMessage::Request { id, request })
}

#[cfg(test)]
impl CoreMessage {
    #[must_use]
    pub fn to_value(&self) -> Value {
        match self {
            Self::Request { id, request } => request.to_value(*id),
            Self::Cancel(p) => notification_value("$/cancel", to_value(p)),
            Self::Response { id, outcome } => response_value(*id, outcome),
        }
    }
}

#[must_use]
pub fn cancel_value(id: u64) -> Value {
    notification_value("$/cancel", json!({ "id": id }))
}

#[must_use]
pub fn response_ok(id: u64, result: Value) -> Value {
    response_value(id, &Ok(result))
}
